src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicSafeHandle.cs

```diff
@@ -52,6 +61,44 @@ protected override bool ReleaseHandle()
         return true;
     }
 
+    public TResult SafeCall<TResult>(Func<MsQuicSafeHandle, TResult> call)
+    {
+        ObjectDisposedException.ThrowIf(IsInvalid, this);
+        bool success = false;
+        try
+        {
+            DangerousAddRef(ref success);
+            Debug.Assert(success);
+            return call(this);
+        }
+        finally
+        {
+            if (success)
+            {
+                DangerousRelease();
+            }
+        }
+    }
+
+    public void SafeCall(Action<MsQuicSafeHandle> call)
```

Comment:
This generally should not be need. runtime would automatically grab reference on pinvoke if we pass in safe handle. .... But we don't. It seems like we lost that when switching to generated API. This is all hidden as we pass the safe handle around everywhere and then we don't use it when we should and we pass IntPtr  to quic.

Reply:
I am not thrilled about the fact that this leads to closure allocation per call in many cases (StreamSend, EnableReceive, ....). I think we should consider (possibly in follow-up) either
- writing wrappers for MsQuicApi which does this for us (i.e. duplicating what P/Invoke does)
- look for a way to have the source generator generate them for us.

Reply:
I agree with both comments, and I believe we should look for a way to solve this in generated API, otherwise it is not sustainable. If we would be able to do that in time, it would be ideal.